### PR TITLE
[ClassLoader] Fix ClassCollectionLoader inlining with declare(strict_types=1)

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Branch? | 2.7 |
| Bug fix? | yes |
| Tests pass? | yes |
| Fixed tickets | #19845 |
| License | MIT |

See diff as https://github.com/symfony/symfony/pull/19859/files?w=1

### CODE REVIEW

**Comment by fabpot** [comment]
So, basically, you prevent files with strict types to be inlined, is that correct?

**Comment by nicolas-grekas** [comment]
I "inline" them by doing a straight "require". We can't exclude them, because they might be parent classes of non strict classes.

File: src/Symfony/Component/ClassLoader/ClassCollectionLoader.php [hunk 1]
```diff
@@ -58,7 +58,12 @@ public static function load($classes, $cacheDir, $name, $autoReload, $adaptive =
 
         $classes = array_unique($classes);
 
-        $cache = $cacheDir.'/'.$name.$extension;
+        // cache the core classes
+        if (!is_dir($cacheDir) && !@mkdir($cacheDir, 0777, true) && !is_dir($cacheDir)) {
+            throw new \RuntimeException(sprintf('Class Collection Loader was not able to create directory "%s"', $cacheDir));
+        }
+        $cacheDir = rtrim(realpath($cacheDir), '/'.DIRECTORY_SEPARATOR);
+        $cache = $cacheDir.DIRECTORY_SEPARATOR.$name.$extension;
```
Comment (fabpot) [inline src/Symfony/Component/ClassLoader/ClassCollectionLoader.php]: This whole section does not look related to the fix.

Comment (nicolas-grekas) [inline src/Symfony/Component/ClassLoader/ClassCollectionLoader.php]: It is: these are some lines move from the bottom to the before the below loop.
https://github.com/symfony/symfony/pull/19859/files?w=1

File: src/Symfony/Component/ClassLoader/ClassCollectionLoader.php [hunk 2]
```diff
@@ -58,7 +58,12 @@ public static function load($classes, $cacheDir, $name, $autoReload, $adaptive =
 
         $classes = array_unique($classes);
 
-        $cache = $cacheDir.'/'.$name.$extension;
+        // cache the core classes
+        if (!is_dir($cacheDir) && !@mkdir($cacheDir, 0777, true) && !is_dir($cacheDir)) {
+            throw new \RuntimeException(sprintf('Class Collection Loader was not able to create directory "%s"', $cacheDir));
```
Comment (HeahDude) [inline src/Symfony/Component/ClassLoader/ClassCollectionLoader.php]: then why not throwing at the very beginning of the method instead?

Comment (nicolas-grekas) [inline src/Symfony/Component/ClassLoader/ClassCollectionLoader.php]: Because it wouldn't work, look at the code before...

Comment (HeahDude) [inline src/Symfony/Component/ClassLoader/ClassCollectionLoader.php]: I've looked at the code, I just didn't get it :)

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.
